packages/core/content-manager/admin/src/pages/EditView/components/FormInputs/Relations/RelationModal.tsx

```diff
@@ -50,47 +66,259 @@ const CustomModalContent = styled(Modal.Content)`
   max-height: 100%;
 `;
 
-const RelationModal = ({ open, onToggle }: RelationModalProps) => {
+type RelationModalWrapperProps = Omit<
+  RelationModalProps,
+  | 'setOnConfirm'
+  | 'setIsConfirmationOpen'
+  | 'isFormModified'
+  | 'open'
+  | 'onToggle'
+  | 'setIsModalOpen'
+>;
+
+const RelationModalWrapper = ({ newDocument, triggerButtonLabel }: RelationModalWrapperProps) => {
+  const navigate = useNavigate();
+  const { pathname, search } = useLocation();
+  const [triggerRefetchDocument] = useLazyGetDocumentQuery();
   const { formatMessage } = useIntl();
+  const documentResponse = useDocumentContext('RelationModalBody', (state) => state.document);
+  const rootDocumentMeta = useDocumentContext(
+    'RelationModalBody',
+    (state) => state.rootDocumentMeta
+  );
+  const documentMeta = useDocumentContext('RelationModalBody', (state) => state.meta);
+  const changeDocument = useDocumentContext('RelationModalBody', (state) => state.changeDocument);
+  const [isConfirmationOpen, setIsConfirmationOpen] = React.useState<boolean>(false);
+  // handler for the confirmation dialog to be used when we confirm the choice
+  const [confirmationDialogPosition, setConfirmationDialogPosition] = React.useState<
+    'cancel' | 'navigate' | 'changeDocument'
+  >();
+  const [isModalOpen, setIsModalOpen] = React.useState(false);
+
+  const initialValues = documentResponse.getInitialFormValues();
+
+  const handleToggleModal = () => {
+    if (isModalOpen) {
+      setIsModalOpen(false);
+      const document = {
+        collectionType: rootDocumentMeta.collectionType,
+        model: rootDocumentMeta.model,
+        documentId: rootDocumentMeta.documentId,
+      };
+      // Change back to the root document
+      changeDocument(document);
+      // Read from cache or refetch root document
+      triggerRefetchDocument(
+        document,
+        // Favor the cache
+        true
+      );
+    } else {
+      setIsModalOpen(true);
+    }
+  };
+
+  const handleCloseModal = (isFormModified: boolean) => {
+    if (isFormModified) {
+      setIsConfirmationOpen(true);
+    } else {
+      handleToggleModal();
+    }
+  };
+
+  const handleChangeModalContent = () => {
+    changeDocument(newDocument);
+  };
+
+  const getFullPageLink = (): string => {
+    const isSingleType = documentMeta.collectionType === SINGLE_TYPES;
+    const queryParams = documentMeta.params?.locale
+      ? `?plugins[i18n][locale]=${documentMeta.params.locale}`
+      : '';
+
+    return `/content-manager/${documentMeta.collectionType}/${documentMeta.model}${isSingleType ? '' : '/' + documentMeta.documentId}${queryParams}`;
+  };
+
+  const editViewUrl = `${pathname}${search}`;
+  const modalDocumentUrlEqualEditViewUrl = editViewUrl.includes(getFullPageLink());
+  const handleRedirection = () => {
+    if (modalDocumentUrlEqualEditViewUrl) {
+      handleToggleModal();
```

Comment:
If these variables are only used in this function scope I would define them in that scope.

```suggestion
  const handleRedirection = () => {
    const editViewUrl = `${pathname}${search}`;
    const isRootDocumentUrl = editViewUrl.includes(getFullPageLink());
    
    if (isRootDocumentUrl) {
      handleToggleModal();
```